### PR TITLE
Prevent reentry of a runner instance

### LINKED ISSUE #6595: nonasyncrunner - KeyError: 6 in threaded test_foreach_dataset.test_basic_resilience
An extract from https://github.com/datalad/datalad/issues/6482#issuecomment-1076484326 which was resolved by a workaround.

was observed while testing the workaround #6587 on windows (although most likely a generic one, but was observed only on windows for now, marking as such).  Brief extract

```
> c:\users\datalad\pycharmprojects\datalad\datalad\runner\nonasyncrunner.py(592)remove_file_number()
-> self.fileno_mapping[file_number],
(Pdb) l
587             """
588
589             # TODO: check exception
590             # Let the protocol know that the connection was lost.
591             self.protocol.pipe_connection_lost(
592  ->             self.fileno_mapping[file_number],
593                 None)
594
595             if file_number in self.active_file_numbers:
596                 # Remove the file number from the set of active numbers.
597                 self.active_file_numbers.remove(file_number)
(Pdb) p https://uchealth.zoom.us/j/514490116
*** SyntaxError: invalid syntax
(Pdb) p self.fileno_mapping[file_number],
*** KeyError: 6
(Pdb) p self.fileno_mapping[file_number]
*** KeyError: 6
(Pdb) p self.fileno_mapping
{3: 1, 4: 2, None: 0}
```

we troubleshooted a bit with @christian-monch  but have not arrived to a conclusion.  So far the idea for the cause is that base process gets threaded so that `ThreadedRunner` instance gets used both in main and thread, causing the `run` to be called twice with different settings which causes this `6` to appear in fileno_mapping but not listed in other places

### PR DESCRIPTION
**Pull request by christian-monch**
This PR addresses issue #6595 

Fixes #6595 

The commits in this PR add a re-entry detection for the methode `ThreadedRunner.run()`, which is not re-entrant. It should also not be called if the protocol that was used is a subclass of `GeneratorMixIn` and the generator returned by `ThreadedRunner.run()` has not yet raised `StopIteration`.

Both cases are detected and handled by this PR. In the non-generator case, the invocations will be serialized. That means the second caller will only be allowed to enter the method if the current caller has left it. In the generator-case, a `RuntimeError` is raised. This might lead to different, new errors, but will provide deterministic detection of the re-entrance issue, that is, of issue #6595

#### 🐛 Bug Fixes
- Prevent reentry of a runner instance. Fixes #6595

### CODE REVIEW

**Review by yarikoptic — CHANGES_REQUESTED** [review]
quick one without going through code yet
- please base-on/target `maint` (would need to use nose construct for exceptions raises, but can keep plain `assert`s imho so it would become 'pytest native' when merged into master) 
- one appveyor run on ubuntu with python 3.7 fail seems to be related:

```
Versions: annexremote=1.6.0 boto=2.49.0 cmd:7z=16.02 cmd:annex=8.20210903 cmd:bundled-git=UNKNOWN cmd:git=2.35.1 cmd:system-git=2.35.1 cmd:system-ssh=8.2p1 datalad=0.16.4+180.ge9aad5b5d duecredit=0.9.1 exifread=3.0.0 humanize=4.1.0 iso8601=1.0.2 keyring=23.5.1 keyrings.alt=UNKNOWN msgpack=1.0.3 mutagen=1.45.1 patoolib=1.12 platformdirs=2.5.2 requests=2.27.1 tqdm=4.64.0
Obscure filename: str=b' |;&%b5{}\'"\xce\x94\xd0\x99\xd7\xa7\xd9\x85\xe0\xb9\x97\xe3\x81\x82 .datc ' repr=' |;&%b5{}\'"ΔЙקم๗あ .datc '
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='UTF-8'
Environment: PATH='/home/appveyor/projects/datalad/__testhome__/../tools/coverage-bin:/home/appveyor/venv3.7.12/bin:/home/appveyor/.rvm/gems/ruby-2.7.2/bin:/home/appveyor/.rvm/gems/ruby-2.7.2@global/bin:/home/appveyor/.rvm/rubies/ruby-2.7.2/bin:/usr/lib/jvm/java-9-openjdk-amd64/bin:/home/appveyor/.gvm/pkgsets/go1.17.7/global/bin:/home/appveyor/.gvm/gos/go1.17.7/bin:/home/appveyor/.gvm/pkgsets/go1.17.7/global/overlay/bin:/home/appveyor/.gvm/bin:/home/appveyor/.gvm/bin:/home/appveyor/.nvm/versions/node/v8.17.0/bin:/opt/octopus:/opt/appveyor/build-agent:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/home/appveyor/.dotnet/tools:/home/appveyor/.rvm/bin:/opt/mssql-tools/bin:/home/appveyor/vcpkg' LANG='C.UTF-8' GIT_CONFIG_PARAMETERS="'init.defaultBranch=dl-test-branch' 'clone.defaultRemoteName=dl-test-remote'" GIT_CONFIG_GLOBAL='/home/appveyor/DLTMP/datalad_temp_a0y9bnr8/.gitconfig' GIT_ASKPASS='true'
=================================================================================================================== FAILURES ===================================================================================================================
________________________________________________________________________________________________________ test_thread_reentry_detection _________________________________________________________________________________________________________
    def test_thread_reentry_detection():
        def run_on(runner: ThreadedRunner,
                   condition: threading.Condition,
                   wait_for_condition: bool):
    
            if wait_for_condition:
                condition.wait()
    
            for _ in runner.run():
                if not wait_for_condition:
                    condition.notify()
                time.sleep(random.random())
    
        # make exceptions visible to test thread
        def new_hook(*args):
            exceptions.append(args[0].exc_type)
    
        threading.excepthook = new_hook
    
        exceptions = []
    
        shared_runner = ThreadedRunner(
            cmd=py2cmd("for i in range(10): print(i)"),
            protocol_class=MinimalGeneratorProtocol,
            stdin=None)
    
        enter_condition = threading.Condition()
        thread_1 = threading.Thread(
            name="thread_1",
            target=run_on,
            args=(shared_runner, enter_condition, False))
    
        thread_2 = threading.Thread(
            name="thread_2",
            target=run_on,
            args=(shared_runner, enter_condition, True))
    
        thread_1.start()
        thread_2.start()
    
        thread_1.join()
        thread_2.join()
    
>       assert exceptions == [RuntimeError]
E       AssertionError: assert [] == [<class 'RuntimeError'>]
E         Right contains one more item: <class 'RuntimeError'>
E         Full diff:
E         - [<class 'RuntimeError'>]
E         + []
../datalad/runner/tests/test_threadsafety.py:75: AssertionError
```

<details>
<summary>and that fail seems to be consistent across different runs/commits</summary> 

```shell
$> git grep 'test_thread_reentry_detection'
02/pr/6737/60b6f94/appveyor-8048-failed/o7ip9u3udu7odsg6.txt:[00:15:51] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection Exception in thread thread_2:
02/pr/6737/60b6f94/appveyor-8048-failed/o7ip9u3udu7odsg6.txt:[00:25:59] ________________________________________________________________________________________________________ test_thread_reentry_detection _________________________________________________________________________________________________________
02/pr/6737/60b6f94/appveyor-8048-failed/o7ip9u3udu7odsg6.txt:[00:25:59]     def test_thread_reentry_detection():
02/pr/6737/60b6f94/appveyor-8048-failed/o7ip9u3udu7odsg6.txt:[00:25:59] FAILED ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection - AssertionError: assert [] == [<class 'RuntimeError'>]
02/pr/6737/60b6f94/appveyor-8048-incomplete/cn5aifgn1xngxklu.txt:[00:28:32] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/60b6f94/appveyor-8048-success/5lfvqyk1yt5ey4b7.txt:[00:15:34] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/60b6f94/travis-14083-success/1.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/60b6f94/travis-14083-success/2.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/60b6f94/travis-14083-success/3.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/60b6f94/travis-14083-success/4.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/60b6f94/travis-14083-success/7.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/60b6f94/travis-14083-success/8.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/98317da/appveyor-8054-failed/n9sv3uw6h20afd5b.txt:[00:16:06] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection Exception in thread thread_2:
02/pr/6737/98317da/appveyor-8054-failed/n9sv3uw6h20afd5b.txt:[00:25:31] ________________________________________________________________________________________________________ test_thread_reentry_detection _________________________________________________________________________________________________________
02/pr/6737/98317da/appveyor-8054-failed/n9sv3uw6h20afd5b.txt:[00:25:31]     def test_thread_reentry_detection():
02/pr/6737/98317da/appveyor-8054-failed/n9sv3uw6h20afd5b.txt:[00:25:31] FAILED ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection - AssertionError: assert [] == [<class 'RuntimeError'>]
02/pr/6737/98317da/appveyor-8054-success/8ptyx0ndb2oe2t6k.txt:[00:16:03] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/98317da/appveyor-8054-success/f27gyim14j4pn1wo.txt:[00:26:31] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/98317da/appveyor-8054-success/l3y8rje1v8gq7sq9.txt:[00:35:13] ..\datalad\runner\tests\test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/98317da/travis-14089-success/1.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/98317da/travis-14089-success/2.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/98317da/travis-14089-success/3.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/98317da/travis-14089-success/4.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/98317da/travis-14089-success/7.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/98317da/travis-14089-success/8.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/abbf61b/appveyor-8049-failed/fla2okr02r6snseh.txt:[00:17:18] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection Exception in thread thread_2:
02/pr/6737/abbf61b/appveyor-8049-failed/fla2okr02r6snseh.txt:[00:26:52] ________________________________________________________________________________________________________ test_thread_reentry_detection _________________________________________________________________________________________________________
02/pr/6737/abbf61b/appveyor-8049-failed/fla2okr02r6snseh.txt:[00:26:52]     def test_thread_reentry_detection():
02/pr/6737/abbf61b/appveyor-8049-failed/fla2okr02r6snseh.txt:[00:26:52] FAILED ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection - AssertionError: assert [] == [<class 'RuntimeError'>]
02/pr/6737/abbf61b/appveyor-8049-success/7nc7m2bcexeocwwl.txt:[00:15:59] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/abbf61b/appveyor-8049-success/x3jbl1n3g7qrb8m9.txt:[00:35:07] ..\datalad\runner\tests\test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/abbf61b/appveyor-8049-success/xaq7go243yl9eql3.txt:[00:29:14] ../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED
02/pr/6737/abbf61b/travis-14084-success/1.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/abbf61b/travis-14084-success/2.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/abbf61b/travis-14084-success/3.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 55%]
02/pr/6737/abbf61b/travis-14084-success/4.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/abbf61b/travis-14084-success/7.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]
02/pr/6737/abbf61b/travis-14084-success/8.txt:../datalad/runner/tests/test_threadsafety.py::test_thread_reentry_detection PASSED [ 53%]

```
</details>

File: datalad/tests/utils.py [hunk 1]
```diff
@@ -273,8 +273,8 @@ def  _wrap_skip_if(*args, **kwargs):
 
 
 def skip_ssh(func):
-    """Skips SSH tests if on windows or if environment variable
-    DATALAD_TESTS_SSH was not set
+    """Skips SSH tests if datalad.tests.ssh (env var DATALAD_TESTS_SSH)
+    is not set at all or set to '0', 'false', or 'no'.
```
Comment (yarikoptic) [inline datalad/tests/utils.py:277]: this and below look unrelated to this PR.  I guess (didn't look) the base might have been `maint` but PR made against `master`?

Comment (christian-monch) [reply datalad/tests/utils.py]: You are right. The branch was created on maint and I forgot to change the PR to be against maint

File: datalad/runner/nonasyncrunner.py [hunk 2]
```diff
@@ -291,6 +303,12 @@ def run(self) -> Union[Any, Generator]:
                # get the return code of the process
                result = gen.return_code
         """
+        self.initialization_lock.acquire()
```
[the remark this answers is not in the document]
Comment (christian-monch) [reply datalad/runner/nonasyncrunner.py:306]: Thx 👍🏻 

Will wrap the locked execution and use a context handler

File: datalad/runner/nonasyncrunner.py [hunk 3]
```diff
@@ -291,6 +303,13 @@ def run(self) -> Union[Any, Generator]:
                # get the return code of the process
                result = gen.return_code
         """
+        with self.initialization_lock:
+            return self._locked_run()
```
Comment (yarikoptic) [inline datalad/runner/nonasyncrunner.py:307]: Fine with me but why not that alternative of @fasteners.locked?

**Comment by yarikoptic** [comment]
Note: appveyor still fails on that single Ubuntu

File: datalad/runner/nonasyncrunner.py [hunk 4]
```diff
@@ -430,9 +450,11 @@ def run(self) -> Union[Any, Generator]:
         self.process_waiting_thread.start()
 
         if issubclass(self.protocol_class, GeneratorMixIn):
-            return _ResultGenerator(self, self.protocol.result_queue)
+            self.generator = _ResultGenerator(self, self.protocol.result_queue)
+            return self.generator
 
-        return self.process_loop()
+        result = self.process_loop()
+        return result
```
[the remark this answers is not in the document]
Comment (christian-monch) [reply datalad/runner/nonasyncrunner.py:457]: Thx!

File: datalad/runner/nonasyncrunner.py (hunk 3 above)
Comment (christian-monch) [reply datalad/runner/nonasyncrunner.py:307]: I guess it is a matter of taste. I find the @fastener.locked-approach with the self._lock attribute less transparent

**Comment by christian-monch** [comment]
> 

Fixed in 80db256432679de0db02776860a8001b99b2b3e8

File: datalad/runner/nonasyncrunner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (christian-monch) [reply datalad/runner/nonasyncrunner.py]: fixed in fd79d2370db43025172b9929010df156e6690ea2

File: datalad/runner/nonasyncrunner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (christian-monch) [reply datalad/runner/nonasyncrunner.py]: fixed in a4519d889478d8faeeafbd71140c40e273fa1453

**Comment by christian-monch** [comment]
The remaining appveyor Mac38 failing test seems unrelated, fails randomly and has mostly passed in this PR

File: datalad/runner/nonasyncrunner.py (hunk 3 above)
Comment (yarikoptic) [reply datalad/runner/nonasyncrunner.py:307]: ok. May be some time we will come with a design doc to unify the tastes on this regard ;)

**Comment by yarikoptic** [comment]
Mac fail -- good old https://github.com/datalad/datalad/issues/6599 (attn @bpoldrack) .

Overall -- great work @christian-monch !  Just left a few comments on a tests fixup commit.

**Comment by christian-monch** [comment]
> [...] good amount of avoidable code duplication [...]

I had it done as suggested earlier, didn't like it the way I did it. Let me do one round of improvement

**Comment by christian-monch** [comment]
@yarikoptic Sorry for the delay, was away for a week.

The tests are refactored and the title is modified according to your suggestions

**Comment by yarikoptic** [comment]
CI is happy, I am happy ;) Thank you @christian-monch !
